java-tests/validating/src/test/kotlin/io/spine/test/options/DistinctITest.kt

```diff
@@ -26,58 +26,184 @@
-import io.spine.base.fieldPath
-import io.spine.protobuf.TypeConverter.toAny
-import io.spine.test.tools.validate.ProtoSet
-import io.spine.test.tools.validate.protoSet
-import io.spine.validate.constraintViolation
+import io.kotest.matchers.maps.shouldContain
+import io.kotest.matchers.shouldBe
+import io.spine.base.FieldPath
+import io.spine.protobuf.Durations2.hoursAndMinutes
+import io.spine.protobuf.Durations2.minutes
+import io.spine.test.tools.validate.UniqueMessageCollections
+import io.spine.test.tools.validate.UniquePrimitiveCollections
+import io.spine.test.tools.validate.uniqueMessageCollections
+import io.spine.test.tools.validate.uniquePrimitiveCollections
+import io.spine.validate.ValidationException
 import org.junit.jupiter.api.DisplayName
+import org.junit.jupiter.api.Nested
 import org.junit.jupiter.api.Test
 import org.junit.jupiter.api.assertDoesNotThrow
+import org.junit.jupiter.api.assertThrows
 
 @DisplayName("`(distinct)` option should be compiled, so that")
 internal class DistinctITest {
 
-    @Test
-    fun `duplicated entries result in a violation`() {
-        val msg = ProtoSet.newBuilder()
-            .addElement(toAny("123"))
-            .addElement(toAny("321"))
-            .addElement(toAny("123"))
-            .buildPartial()
-
-        val error = msg.validate()
-        error.shouldBePresent()
-
-        val violations = error.get().constraintViolationList
-        val expected = constraintViolation {
-            fieldPath = fieldPath { fieldName.add("element") }
+    @Nested inner class
+    `when handling 'list' field` {
+
+        @Test
+        fun `duplicated primitive entries result in a violation`() {
+            val duplicate1 = 123
+            val duplicate2 = 321
+
+            val exception = assertThrows<ValidationException> {
+                uniquePrimitiveCollections {
+                    element.add(duplicate1)
+                    element.add(1)
+                    element.add(duplicate1)
+                    element.add(2)
+                    element.add(duplicate2)
+                    element.add(3)
+                    element.add(duplicate2)
+                }
+            }
+
+            val violation = exception.constraintViolations.first()
+            val duplicates = listOf(duplicate1, duplicate2)
+            with(violation) {
+                fieldPath shouldBe FieldPath("element")
+                message.placeholderValueMap shouldContain ("field.duplicates" to "$duplicates")
+            }
         }
 
-        assertThat(violations)
-            .comparingExpectedFieldsOnly()
-            .containsExactly(expected)
+        @Test
+        fun `duplicated message entries result in a violation`() {
+            val duplicate1 = hoursAndMinutes(2, 20)
+            val duplicate2 = hoursAndMinutes(18, 40)
+
+            val exception = assertThrows<ValidationException> {
+                uniqueMessageCollections {
+                    element.add(duplicate1)
```

Comment:
Similar question on `element.run { ... }` here. Just to same some eye movements for the sake of readers who struggle not to start wearing glasses.